File: problems/032-conflict-package-cover/solutions/c/main.c
#include <stdio.h>
#include <stdlib.h>
int main(void) {
  int L, R, M;
  if (scanf("%d%d%d", &L, &R, &M) != 3)
    return 0;
  int *head = malloc(sizeof(int) * (size_t)L),
      *to = malloc(sizeof(int) * (size_t)(M ? M : 1)),
      *next = malloc(sizeof(int) * (size_t)(M ? M : 1));
  for (int i = 0; i < L; i++)
    head[i] = -1;
  for (int e = 0; e < M; e++) {
    int u, v;
    scanf("%d%d", &u, &v);
    u--;
    to[e] = v - 1;
    next[e] = head[u];
    head[u] = e;
  }
  int *pu = malloc(sizeof(int) * (size_t)L),
      *pv = malloc(sizeof(int) * (size_t)R),
      *dist = malloc(sizeof(int) * (size_t)L),
      *q = malloc(sizeof(int) * (size_t)L),
      *cur = malloc(sizeof(int) * (size_t)L),
      *su = malloc(sizeof(int) * (size_t)L),
      *sv = malloc(sizeof(int) * (size_t)L);
  for (int i = 0; i < L; i++)
    pu[i] = -1;
  for (int i = 0; i < R; i++)
    pv[i] = -1;
  int matching = 0;
  for (;;) {
    int l = 0, r = 0, terminal = -1;
    for (int u = 0; u < L; u++) {
      dist[u] = pu[u] < 0 ? 0 : -1;
      if (pu[u] < 0)
        q[r++] = u;
    }
    while (l < r) {
      int u = q[l++];
      if (terminal >= 0 && dist[u] >= terminal)
        continue;
      for (int e = head[u]; e >= 0; e = next[e]) {
        int w = pv[to[e]];
        if (w < 0)
          terminal = dist[u];
        else if (dist[w] < 0) {
          dist[w] = dist[u] + 1;
          q[r++] = w;
        }
      }
    }
    if (terminal < 0)
      break;
    for (int i = 0; i < L; i++)
      cur[i] = head[i];
    for (int root = 0; root < L; root++) {
      if (pu[root] >= 0)
        continue;
      int ut = 1, vt = 0, ok = 0;
      su[0] = root;
      while (ut && !ok) {
        int u = su[ut - 1], desc = 0;
        while (cur[u] >= 0) {
          int e = cur[u];
          cur[u] = next[e];
          int v = to[e], w = pv[v];
          if (w < 0 && dist[u] == terminal) {
            pu[u] = v;
            pv[v] = u;
            for (int i = vt - 1; i >= 0; i--) {
              pu[su[i]] = sv[i];
              pv[sv[i]] = su[i];
            }
            ok = 1;
            break;
          }
          if (w >= 0 && dist[u] < terminal && dist[w] == dist[u] + 1) {
            sv[vt++] = v;
            su[ut++] = w;
            desc = 1;
            break;
          }
        }
        if (!ok && !desc) {
          dist[u] = -1;
          ut--;
          if (vt)
            vt--;
        }
      }
      if (ok)
        matching++;
    }
  }
  printf("%d\n", matching);
  return 0;
}

File: problems/032-conflict-package-cover/solutions/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	var L, R, M int
	fmt.Fscan(in, &L, &R, &M)
	g := make([][]int, L)
	for ; M > 0; M-- {
		var u, v int
		fmt.Fscan(in, &u, &v)
		g[u-1] = append(g[u-1], v-1)
	}
	pu := make([]int, L)
	pv := make([]int, R)
	dist := make([]int, L)
	for i := range pu {
		pu[i] = -1
	}
	for i := range pv {
		pv[i] = -1
	}
	matching := 0
	for {
		q := []int{}
		for u := 0; u < L; u++ {
			if pu[u] < 0 {
				dist[u] = 0
				q = append(q, u)
			} else {
				dist[u] = -1
			}
		}
		terminal := -1
		for h := 0; h < len(q); h++ {
			u := q[h]
			if terminal >= 0 && dist[u] >= terminal {
				continue
			}
			for _, v := range g[u] {
				w := pv[v]
				if w < 0 {
					terminal = dist[u]
				} else if dist[w] < 0 {
					dist[w] = dist[u] + 1
					q = append(q, w)
				}
			}
		}
		if terminal < 0 {
			break
		}
		cur := make([]int, L)
		for root := 0; root < L; root++ {
			if pu[root] >= 0 {
				continue
			}
			su, sv := []int{root}, []int{}
			ok := false
			for len(su) > 0 && !ok {
				u := su[len(su)-1]
				down := false
				for cur[u] < len(g[u]) {
					v := g[u][cur[u]]
					cur[u]++
					w := pv[v]
					if w < 0 && dist[u] == terminal {
						pu[u] = v
						pv[v] = u
						for i := len(sv) - 1; i >= 0; i-- {
							pu[su[i]] = sv[i]
							pv[sv[i]] = su[i]
						}
						ok = true
						break
					}
					if w >= 0 && dist[u] < terminal && dist[w] == dist[u]+1 {
						sv = append(sv, v)
						su = append(su, w)
						down = true
						break
					}
				}
				if !ok && !down {
					dist[u] = -1
					su = su[:len(su)-1]
					if len(sv) > 0 {
						sv = sv[:len(sv)-1]
					}
				}
			}
			if ok {
				matching++
			}
		}
	}
	fmt.Println(matching)
}
